api/grpcserver/transaction_service.go

```diff
@@ -0,0 +1,267 @@
+package grpcserver
+
+import (
+	"bytes"
+	pb "github.com/spacemeshos/api/release/go/spacemesh/v1"
+	"github.com/spacemeshos/go-spacemesh/api"
+	"github.com/spacemeshos/go-spacemesh/common/types"
+	"github.com/spacemeshos/go-spacemesh/events"
+	"github.com/spacemeshos/go-spacemesh/log"
+	"github.com/spacemeshos/go-spacemesh/miner"
+	"github.com/spacemeshos/go-spacemesh/p2p/peers"
+	"golang.org/x/net/context"
+	"google.golang.org/genproto/googleapis/rpc/code"
+	rpcstatus "google.golang.org/genproto/googleapis/rpc/status"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
+)
+
+// TransactionService exposes transaction data, and a submit tx endpoint
+type TransactionService struct {
+	Network     api.NetworkAPI // P2P Swarm
+	Mesh        api.TxAPI      // Mesh
+	Mempool     *miner.TxMempool
+	GenTime     api.GenesisTimeAPI
+	PeerCounter api.PeerCounter
+	Syncer      api.Syncer
+}
+
+// RegisterService registers this service with a grpc server instance
+func (s TransactionService) RegisterService(server *Server) {
+	pb.RegisterTransactionServiceServer(server.GrpcServer, s)
+}
+
+// NewGlobalStateService creates a new grpc service using config data.
+func NewTransactionService(
+	net api.NetworkAPI, tx api.TxAPI, mempool *miner.TxMempool, genTime api.GenesisTimeAPI,
+	syncer api.Syncer) *TransactionService {
+	return &TransactionService{
+		Network:     net,
+		Mesh:        tx,
+		Mempool:     mempool,
+		GenTime:     genTime,
+		PeerCounter: peers.NewPeers(net, log.NewDefault("grpc_server.TransactionService")),
+		Syncer:      syncer,
+	}
+}
+
+// SubmitTransaction allows a new tx to be submitted
+func (s TransactionService) SubmitTransaction(ctx context.Context, in *pb.SubmitTransactionRequest) (*pb.SubmitTransactionResponse, error) {
+	log.Info("GRPC TransactionService.SubmitTransaction")
+
+	tx, err := types.BytesToTransaction(in.Transaction)
+	if err != nil {
+		log.Error("failed to deserialize tx, error: %v", err)
+		return nil, status.Errorf(codes.InvalidArgument,
+			"`Transaction` must contain a valid, serialized transaction")
+	}
+	log.Info("GRPC TransactionService.SubmitTransaction to address: %s (len: %v), "+
```

Comment:
do we use gas limit?

Reply:
The transaction type has a field called `GasLimit`. Not sure if it's currently used or not. It's implemented in the existing API, and used in tests, etc.